File: core/src/world/chunk/position.rs
use crate::world::chunk::CHUNK_SIZE;
use crate::world::position::WorldPos;
use num_integer::Integer;
use std::convert::TryInto;

/// A point representing a chunk position in (x, y) coordinate
/// space. It is computed by dividing WorldPos by [CHUNK_SIZE] and
/// rounding towards negative infinity.
#[derive(Eq, PartialEq, Clone, Copy, Hash, Debug)]
pub struct ChunkPos {
    pub x: i32,
    pub y: i32
}

impl ChunkPos {
    pub fn map_x(self, f: impl Fn(i32) -> i32) -> Self {
        Self {
            x: f(self.x),
            y: self.y,
        }
    }

    pub fn map_y(self, f: impl Fn(i32) -> i32) -> Self {
        Self {
            x: self.x,
            y: f(self.y),
        }
    }


    /// [WorldPos] does not implement From<ChunkPos> because
    /// [ChunkPos] has no `z` component.
    pub fn into_wpos(self, z: i8) -> WorldPos {
        let chunk_size: i32 = CHUNK_SIZE.try_into().unwrap();
        WorldPos {
            x: self.x * chunk_size,
            y: self.y * chunk_size,
            z: z
        }
    }
}

impl Default for ChunkPos {
    fn default() -> Self {
        Self {
            x: 0,
            y: 0
        }
    }
}

impl From<WorldPos> for ChunkPos {
    fn from(wpos: WorldPos) -> Self {
        let chunk_size: i32 = CHUNK_SIZE.try_into().unwrap();
        Self {
            x: wpos.x.div_floor(&chunk_size),
            y: wpos.y.div_floor(&chunk_size)
        }
    }
}

File: core/Game/AcidRain/World/Chunk/Position.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE UnicodeSyntax #-}
module Game.AcidRain.World.Chunk.Position
  ( ChunkPos(..), cpX, cpY
  , toWorldPos
  ) where

import Data.Convertible.Base (Convertible(..))
import Data.Hashable (Hashable)
import Data.Int (Int8, Int32)
import Game.AcidRain.World.Position (WorldPos(..), wpX, wpY)
import Game.AcidRain.World.Chunk (chunkSize)
import GHC.Generics (Generic)
import Lens.Micro ((^.))
import Lens.Micro.TH (makeLenses)
import Prelude.Unicode ((⋅))


-- | A point representing a chunk position in (x, y) coordinate
-- space. It is computed by dividing 'WorldPos' by 'chunkSize' and
-- rounding towards negative infinity.
data ChunkPos = ChunkPos
  { _cpX ∷ {-# UNPACK #-} !Int32
  , _cpY ∷ {-# UNPACK #-} !Int32
  } deriving (Eq, Show, Generic)

makeLenses ''ChunkPos

instance Hashable ChunkPos

instance Convertible WorldPos ChunkPos where
  safeConvert wp
    = Right $ ChunkPos
      { _cpX = wp^.wpX `div` chunkSize
      , _cpY = wp^.wpY `div` chunkSize
      }

-- | There is no instance of @'Convertible' 'WorldPos' 'ChunkPos'@
-- because 'ChunkPos' has no @z@ component.
toWorldPos ∷ ChunkPos → Int8 → WorldPos
toWorldPos cp z
  = WorldPos
    { _wpX = cp^.cpX ⋅ chunkSize
    , _wpY = cp^.cpY ⋅ chunkSize
    , _wpZ = z
    }
